server/src/main/java/io/crate/types/DataType.java

```diff
@@ -174,8 +174,32 @@ public List<DataType<?>> getTypeParameters() {
      * @param other The other type to compare against.
      * @return True if the current type precedes, false otherwise.
      */
-    public boolean precedes(DataType<?> other) {
-        return this.precedence().ordinal() > other.precedence().ordinal();
+    public boolean precedes(DataType<?> that) {
+        int thisOrdinal = this.precedence().ordinal();
+        int thatOrdinal = that.precedence().ordinal();
+        if (thisOrdinal == thatOrdinal) {
+            Integer thisPrecision = this.numericPrecision();
+            Integer thatPrecision = that.numericPrecision();
+            if (thisPrecision == null && thatPrecision == null) {
+                Integer thisLength = this.characterMaximumLength();
+                Integer thatLength = that.characterMaximumLength();
+                if (thisLength == null) {
+                    return true;
+                }
+                if (thatLength == null) {
+                    return false;
+                }
+                return thisLength > thatLength;
+            }
+            if (thisPrecision == null) {
+                return true;
+            }
+            if (thatPrecision == null) {
+                return false;
+            }
+            return thisPrecision > thatPrecision;
+        }
+        return thisOrdinal > thatOrdinal;
```

Comment:
The changes became much larger compared to https://github.com/crate/crate/pull/16731/commits/894e4cc34e70a22d1f1a4e3bd1c655116b403c73#diff-875fdfcfc352a1b306aa9776e81bd6e1999aa0856a550f043ab9d3164e5ca2f8L193-R196.